drivers/spi/spi_shell.c

```diff
@@ -4,54 +4,278 @@
+static size_t num_spi_shell_devices =
+	ARRAY_SIZE(spi_shell_devices) - CONFIG_SPI_SHELL_NUM_EMPTY_INSTANCES;
+
+#define INST_MAPS_FROM_SPI_DEVICE_NODELABELS(node_id)                                              \
+	RUN_FN_ON_SPI_DEVICE(node_id, DT_FOREACH_NODELABEL_VARGS(node_id, INST_MAP, node_id))
+
+#define INST_MAPS_FROM_SPI_DEVICE_NODE_ID(node_id)                                                 \
+	RUN_FN_ON_SPI_DEVICE(node_id, INST_MAP_FROM_NODE_ID(node_id))
+
+/* A list of maps. At compile time we create maps for all nodelabels and node_ids of spi devices.
+ * Additional empty space is reserved for the spi buses, which are added in spi_buses_init at
+ * runtime.
+ */
+static struct map maps[] = {
+	DT_FOREACH_STATUS_OKAY_NODE(INST_MAPS_FROM_SPI_DEVICE_NODELABELS)
+		DT_FOREACH_STATUS_OKAY_NODE(INST_MAPS_FROM_SPI_DEVICE_NODE_ID)
+			CREATE_NUM_EMPTY_STRUCTS(CONFIG_SPI_SHELL_NUM_EMPTY_INSTANCES)};
+static size_t num_maps = ARRAY_SIZE(maps) - CONFIG_SPI_SHELL_NUM_EMPTY_INSTANCES;
 
 static bool device_is_spi(const struct device *dev)
 {
 	return DEVICE_API_IS(spi, dev);
 }
 
-static void device_name_get(size_t idx, struct shell_static_entry *entry)
+static bool device_is_gpio(const struct device *dev)
+{
+	return DEVICE_API_IS(gpio, dev);
+}
+
+/**
+ * @brief Initialize spi buses at runtime.
+ *
+ * Since Zephyr currently doesn't support getting a device for all spi buses at compile time in a
+ * generic way, we do it at runtime.
+ *
+ * For each spi bus device we:
+ * - add an entry to spi_shell_devices array
+ * - add an entry to maps array by its name
+ * - add an entry to maps array for each it's nodelabel
+ */
+static int spi_buses_init(void)
+{
+	int idx = 0;
+
+	while (1) {
+		const struct device *dev = shell_device_filter(idx, device_is_spi);
+
+		idx++;
+
+		if (dev == NULL) {
+			break;
+		}
+
+		if (num_spi_shell_devices == ARRAY_SIZE(spi_shell_devices)) {
+			printk("ERROR: not enough space in spi_shell_devices array\n");
+			printk("Increase CONFIG_SPI_SHELL_NUM_EMPTY_INSTANCES.\n");
+			break;
+		}
+
+		spi_shell_devices[num_spi_shell_devices++] =
+			INST_SPI_SHELL_DEVICE_AS_SPI_DEV_AS_SPI_BUS(dev);
+
+		maps[num_maps++] = (struct map){
+			.label = dev->name,
+			.name = dev->name,
+		};
+
+		const struct device_dt_nodelabels *nl = device_get_dt_nodelabels(dev);
+
+		if (nl == NULL) {
+			/* No nodelabel for this device, so we can skip the rest. */
+			continue;
+		}
+
+		if (num_maps + nl->num_nodelabels > ARRAY_SIZE(maps)) {
+			printk("ERROR: not enough space in maps array\n");
+			printk("Increase CONFIG_SPI_SHELL_NUM_EMPTY_INSTANCES.\n");
+			break;
+		}
+
+		for (size_t i = 0; i < nl->num_nodelabels; i++) {
+			maps[num_maps++] = (struct map){
+				.label = nl->nodelabels[i],
+				.name = dev->name,
+			};
+		}
+	}
+
+	if (num_spi_shell_devices == 0) {
+		printk("ERROR: no spi devices or spi buses are enabled, check devicetree.\n");
+	}
+
+	return 0;
+}
+
+SYS_INIT(spi_buses_init, APPLICATION, 0);
```

Comment:
Don't use SYS_INIT() for this.

In find_spec_by_label for instance you could trigger the initialization once if some init variable is still false. Set it to true in spi_buses_init, so subsequent call will not re-init again.

Reply:
SYS_INIT() does what you say, but it should be used only for software logic that is necessary to run asap (depending on the level/prio though). Your shell feature for instance, maybe the user will not use it, so until it is being used: there is no need for it to be initialized. 
This macro has be abused in so many ways. You case is by far not the worst. Some use it to initialize only once some part of a multi-instance device driver while a simple static variable acting as a gatekeeper could do the trick for instance.